### PR TITLE
sql: do everything in a transaction

### PR DESCRIPTION
**Pull request by tamird**
This is a pretty crude solution to #1889. I'd like to add some tests before merging this, but I thought I'd solicit opinions before proceeding.

### CODE REVIEW

**Comment by vivekmenezes** [comment]
Tamir,

We need to be able to run commands outside of a transaction. I think what
might make sense is for you to pass in a runner to the planner and have the
planner use the runner to execute the commands.

I'm in the process of creating a change that will marshall/unmarshall
proto.Transaction objects for use across statements. If a transaction is
indeed created then the transaction can be passed in as the runner. The
default implementation is to pass in the db runner.

Vivek

**Comment by tamird** [comment]
> We need to be able to run commands outside of a transaction.

Why? Under what circumstance? The only case I can think of is schema lookup; doing that outside a transaction is not excluded by this change.

Also, a runner (as currently implemented) is far too restrictive, since it can only execute batches. We can widen that interface or introduce a new one, but it won't really change the spirit of this implementation.

**Comment by vivekmenezes** [comment]
In SQL you do have to explicitly specify that you are using a transaction.

Is the runner restrictive for this particular change?

**Comment by bdarnell** [comment]
You don't have to explicitly state that you want a transaction in SQL. If you do not explicitly begin a transaction, SQL uses autocommit mode, which is defined to be equivalent to wrapping each statement in a transaction. More importantly, for our sql implementation, a SQL statement outside a transaction still needs to be a transaction at the KV layer.

Certain commands can be "outside" the transactional system (e.g. schema changes in MySQL), but that's another matter. I think for our purposes we'd always want to use a transaction, and then for things like schema changes they'd trigger actions to be taken outside their transactional scope.

**Comment by vivekmenezes** [comment]
Sounds fine.

Tamir, can you make this to be reflective of what would happen if a
transaction object were not supplied in the request. I.e. req.Txn == nil

Thanks

**Comment by tamird** [comment]
I don't follow; there is no req.Txn, is there?

**Comment by vivekmenezes** [comment]
It's fine the way it is. I can always make this auto commit explicitly in
another PR.

File: sql/plan.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ import (
 // planner is the centerpiece of SQL statement execution combining session
 // state and database state with the logic for SQL execution.
 type planner struct {
-	db      *client.DB
+	db      *client.Txn
```
Comment (vivekmenezes) [inline sql/plan.go]: s/db/txn/g

File: sql/scan.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ type qvalMap map[structured.ColumnID]*parser.ParenExpr
 // A scanNode handles scanning over the key/value pairs for a table and
 // reconstructing them into rows.
 type scanNode struct {
-	db         *client.DB
+	db         *client.Txn
```
Comment (vivekmenezes) [inline sql/scan.go]: same

**Comment by vivekmenezes** [comment]
LGTM

**Comment by petermattis** [comment]
LGTM (this is what I thought was going to be done).

**Comment by vivekmenezes** [comment]
I'm separately working on the cross sql statement transactions.

**Comment by tschottdorf** [comment]
> Also, a runner (as currently implemented) is far too restrictive, since it can only execute batches. We can widen that interface or introduce a new one, but it won't really change the spirit of this implementation.

what's wrong with only doing batches?

Unrelated future consideration:
At some point it's going to be preferable if nodes don't execute their batches but pass control over that up in some form to allow for more parallelism. For example, getting the next chunk from two tables (for example in a JOIN) shouldn't query first one, then the other. Instead, there should only be one batch.

LGTM

**Comment by tamird** [comment]
@tschottdorf nothing wrong with only doing batches, except that it's infeasible so long as e.g. update depends on `select` - that is, if you need to read anything at all, you can't use a batch for _all_ your work.

**Comment by tamird** [comment]
Added tests. Could someone give this another once-over?

File: sql/scan.go (hunk 2 above)
Comment (tamird) [inline sql/scan.go]: done

Comment (tamird) [inline sql/scan.go]: err, no it's not; doing now

**Comment by MBerhault** [comment]
LGTM

File: sql/driver/driver_test.go [hunk 3]
```diff
@@ -223,7 +223,7 @@ func TestInsecure(t *testing.T) {
 	defer func() {
 		_ = db.Close()
 	}()
-	if _, err := db.Exec("CREATE DATABASE t"); err != nil {
+	if _, err := db.Exec(`SELECT 1`); err != nil {
```
Comment (vivekmenezes) [inline sql/driver/driver_test.go]: why did you make this change?

Comment (tamird) [inline sql/driver/driver_test.go]: for simplicity. we're just testing that we can talk to the server in insecure mode. cc @MBerhault

File: sql/insert_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+	descKey := gr.ValueBytes()
+	desc := structured.TableDescriptor{}
+	if err := kvDB.GetProto(descKey, &desc); err != nil {
+		t.Fatal(err)
+	}
+
+	var tablePrefix []byte
+	tablePrefix = append(tablePrefix, keys.TableDataPrefix...)
+	tablePrefix = encoding.EncodeUvarint(tablePrefix, uint64(desc.ID))
```
Comment (vivekmenezes) [inline sql/insert_test.go]: is desc.ID == structured.MaxReservedDescID+1?  In which case you can just use that and get rid of the rest of the code above.

File: sql/driver/driver_test.go (hunk 3 above)
Comment (MBerhault) [inline sql/driver/driver_test.go]: yeah. I really wanted DB.Ping(), but that doesn't necessarily open a connection. Any statement that makes it to the server will do (and they all do).

File: sql/insert_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		t.Fatalf("unexpected error %s", err)
+	}
+
+	// Verify nothing was written.
+	if kvs, err := kvDB.Scan(tableStartKey, tableEndKey, 0); err != nil {
+		t.Fatal(err)
+	} else if l := 12; len(kvs) != l {
+		t.Fatalf("expected %d key value pairs, but got %d", l, len(kvs))
+	}
+}
```
Comment (vivekmenezes) [inline sql/insert_test.go]: BTW Pete's EXPLAIN DEBUG is going to allow us to write these kinds of tests

**Comment by vivekmenezes** [comment]
LGTM

**Comment by vivekmenezes** [comment]
And a big thank you for adding all the tests!

File: sql/insert_test.go (hunk 4 above)
Comment (tamird) [inline sql/insert_test.go]: Ah, no. `desc.ID` is the ID of the table descriptor. `structured.MaxReservedDescID+1` is the ID of the database descripotor.

File: sql/update_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	if _, err := sqlDB.Exec(`UPDATE t.kv SET v = 'd' WHERE k = 'c'`); !testutils.IsError(err, "duplicate key value .* violates unique constraint") {
+		t.Fatalf("unexpected error %s", err)
+	}
+
+	// Verify nothing was written.
+	if kvs, err := kvDB.Scan(tableStartKey, tableEndKey, 0); err != nil {
+		t.Fatal(err)
+	} else if l := 12; len(kvs) != l {
+		t.Fatalf("expected %d key value pairs, but got %d", l, len(kvs))
+	}
```
Comment (petermattis) [inline sql/update_test.go]: What are these new tests testing that the logic tests aren't covering? As far as I can tell they are verifying the precise number of keys that should be present with the table prefix, but they aren't verifying the actual contents of those keys.

Comment (tamird) [inline sql/update_test.go]: Correct, they verifying that these commands are all-or-nothing. They do not
test the internal representation.

**Pull request merged**